tasks/custom_task/custom_task.py

```diff
@@ -0,0 +1,81 @@
+"""
+This module defines the InvokeLogger context manager to log the invoke task information to the DD_INVOKE_LOGS_PATH.
+This will then be uploaded to Datadog's backend with a correct Log Agent configuration.
+"""
+
+from time import perf_counter
+from getpass import getuser
+from datetime import datetime
+import logging
+import traceback
+from invoke import Context
+
+DD_INVOKE_LOGS_PATH = "/tmp/dd_invoke.log"
+
+
+def log_invoke_task(name: str, module: str, task_datetime: str, duration: float, task_result: str) -> None:
+    """
+    Logs the task information to the DD_INVOKE_LOGS_PATH file.
+    This should be uploaded to Datadog's backend with a correct Log Agent configuration:
+    ```
+    logs:
+    - type: file
+        path: "/tmp/dd_invoke.log"
+        service: "dd_invoke_logs"
+        source: "invoke"
+    ```
+    """
+    logging.basicConfig(filename=DD_INVOKE_LOGS_PATH, level=logging.INFO, format='%(message)s')
+    user = getuser()
+    task_info = {
+        "name": name,
+        "module": module,
+        "datetime": task_datetime,
+        "duration": duration,
+        "user": user,
+        "result": task_result,
+    }
+    logging.info(task_info)
+
+
+class InvokeLogger:
+    """
+    Context manager to log an invoke task run information.
+    """
+
+    def __init__(self, task) -> None:
+        self.task = task
+        self.datetime = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
+        self.start = None
+        self.end = None
+
+    def __enter__(self):
+        self.start = perf_counter()
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        self.end = perf_counter()
+        duration = round(self.end - self.start, 4)
+        name = self.task.__name__
+        module = self.task.__module__.replace("tasks.", "")
+        task_result = (
+            None if exc_type is None else "".join(traceback.format_exception(exc_type, exc_value, exc_traceback))
+        )
+        log_invoke_task(
+            name=name, module=module, task_datetime=self.datetime, duration=duration, task_result=task_result
+        )
+
+
+def custom__call__(self, *args, **kwargs):
+    """
+    Custom __call__ method for the Task class.
+    The code was adapted from the invoke 2.2.0 library's Task class.
+    """
+    # Guard against calling tasks with no context.
```

Comment:
We have a lot of tasks calling without context (eg [here](https://github.com/DataDog/datadog-agent/blob/main/tasks/junit_tasks.py#L7)). So I'm worried we won't be able to use these anymore?

Reply:
Could be confusing, but this is just the legacy code, this is invoke code that I copy pasted, I didn't change anything. The only thing I changed is line 78 i.e `with InvokeLogger(self):`